### DISCUSSION TITLE
Returning variables from FreeHandCutPlotter parent class

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#629] [Q&A] [answered]
Hi there Marco!!
So I have finally made it onto that selection window question I asked, and am trying to get my head around the workings of the cut plotter. I decided to make a new thread about it, I think thats a good idea. 
I have decided to extend the FreeHandCutPlotter class, rather than write something from scratch because similar functionality to that which I need is already in there (once again, thanks for an awesome job you've done with this library, its so good!!) However, I am running into problems getting the values from the super class back for examination and was hoping you might have some insight as  to what I am doing wrong here. 

I need to understand exactly how the class segments the points based on the user interaction, so that I can extend it by adding methods for cutters of different geometries. This is what I have so far:

 ```
 class newCutPlotter(vedo.applications.FreeHandCutPlotter):
      """Test for the creation of a class extension for cutting point clouds
      """

      def __init__(self, mesh, **options):
          super().__init__(mesh, **options)


      def newFunction(self, *args, **kwargs):
          """Function that is new!!!
          """
          super().start(*args, **kwargs)

          print("Skip all the way home")


      def getPointsCallback(self, evt):
          """Call back to get the points under the cursor
          """
          if not evt.actor:
              return
          print(f"These are the cpoints: {self.cpoints}")



      def getSplineValsCallback(self, evt):
          """Call back to get the points in the spline
          """

          tol = self.mesh.diagonalSize()/2            # size of ribbon (not shown)
          pts = self.spline.points()
          n = fitPlane(pts, signed=True).normal       # compute normal vector to points
          rb = Ribbon(pts - tol*n, pts + tol*n, closed=True)
          #self.mesh.cutWithMesh(rb, invert=inv)       # CUT

          print(f"These are the tolerance points: {tol}")
          print(f"These are the spline points: {pts}")



  vedo.settings.useParallelProjection = True  # to avoid perspective artifacts

  pts = np.random.randn(10000, 3)
  pts = Points(pts).ps(8)
  pts.name = "My Gaussian"

  cube = Cube().x(2)

  plt = newCutPlotter(pts).addHoverLegend()
  plt.addCallback('RightButtonRelease', plt.getPointsCallback)
  plt.addCallback('KeyPress', plt.getSplineValsCallback)
  plt.newFunction(pts, cube, bg2='lightblue',axes=1)


  print('Done')
  print('Done')

```
This code produces the following error:
```
  File "D:\installs\programs\Anaconda\Anaconda3\envs\nnse\lib\site-packages\vedo\plotter.py", line 1708, in _func_wrap
    func(event_dict)

  File "d:\life\canada\uwo\ece\phd\research\code\nnse\testo3d.py", line 1168, in getSplineValsCallback
    pts = self.spline.points()

AttributeError: 'NoneType' object has no attribute 'points'
```

I am confused by this. How can self.spline be None, while self.mesh has a value? If I remove getSplineValsCallback(), then the cutter runs with full functionality, so I don't think there is a problem with my extension of the class. Maybe you have some insights as to what I am doing wrong here?

Cheers
j

**Comment by marcomusy** [comment, marked as answer]
Hi @jkissi
try
`plt.addCallback('KeyPress', plt.getSplineValsCallback, priority=100)`
so the object is not erased by the previous `z` keypress call of the parent.
Also 
`super().start(*args, **kwargs)`
looks a bit weird to my eye.. should be better
`self.start(*args, **kwargs)`

**Reply by jkissi** [reply]
Thanks Marco, that worked like a charm! Brilliant. 
Am I right in thinking that the self.spline points are the points that should remain after the cutting action?
And also that self.cpoints are the coordinates of the points that make up the cutting line?

**Reply by marcomusy** [reply]
> Am I right in thinking that the self.spline points are the points that should remain after the cutting action?

uhm no, those are the points of the spline (upsampled wrt cpoints if i remember)..
the pts Mesh/Points object is cut in-place

> And also that self.cpoints are the coordinates of the points that make up the cutting line?

Yes

Btw note that you can add a `plt.start()`at the end of the script to repeat the cutting.

**Reply by jkissi** [reply]
so are you saying that the coordinates of the points that are to be removed (or remain) are not stored in any variables in this class? I am asking because that is something I will need to do for my extension....

**Reply by marcomusy** [reply]
I'm afraid not! It's just cut and gone... you would like to mark the points instead of removing them?

Option 1
You may rewrite just the keypress call :
`plt.removeCallback(plt.idkeypress)`

and write a custom version where instead of cutWithMesh() you use `insidePoints()` and then `.threshold()`
analogous to [this](https://github.com/marcomusy/vedo/blob/cdc1d8cad8e56280d019ca62c7555c9bf936302a/vedo/applications.py#L849)

Option 2
I can easily change those lines but you'll need to wait for the next release.

**Reply by jkissi** [reply]
Thanks for the offer Marco but I cannot wait for the next release, so I will attempt to extend it with your suggestion. I will fiddle for a bit and get back to you...hope you're having a great day!

j

**Reply by jkissi** [reply]
Actually, let me _rephrase_ that: I think it's a great idea to do that for the next release Marco but I will also be looking at your first suggestion in the mean time ;-)

**Comment by jkissi** [comment]
Hi there Marco, 
I have been messing around with your suggestions and am having some success. I just wanted to ask: precisely why I am using the threshold() function? Using the insidepoints() function on the ribbon object gives the points....what added functionality does threshold() provide?

**Reply by marcomusy** [reply]
Ouch I'm afraid I've found a small bug in the insidepoints() function..
And - you are right - you dont' really need the `threshold()`

After
`pip install -U git+https://github.com/marcomusy/vedo.git`

you should be able to reproduce this:
```python
from vedo import *

settings.useDepthPeeling = True

pts = Points(np.random.randn(10000, 3)*[3,2,1] + [50,60,70])

elli = pcaEllipsoid(pts, pvalue=0.50)

elli.insidePoints(pts, returnIds=True)

ids  = elli.insidePoints(pts, returnIds=True)
pts.print()  # a new "IsInside" array now exists in pts
pin = pts.points()[ids]
print("inside  points #", len(pin))

# Create an inverted mask instead of calling insidePoints(invert=True)
mask = np.ones(pts.N(), dtype=bool)
mask[ids] = False
pout = pts.points()[mask]
print("outside  points #", len(pout))

show(elli,
     Points(pin).c("green4"),
     Points(pout).c("red5").alpha(0.2),
     axes=1,
).close()
```
![Screenshot from 2022-04-23 01-31-46](https://user-images.githubusercontent.com/32848391/164817324-0d8d48f5-262d-445d-babe-3b0be9371bb9.png)

**Reply by jkissi** [reply]
great, thats awesome!!
